pytorch_lightning/metrics/functional/regression.py

```diff
@@ -182,3 +184,116 @@ def psnr(
     psnr_base_e = 2 * torch.log(data_range) - torch.log(mse_score)
     psnr = psnr_base_e * (10 / torch.log(torch.tensor(base)))
     return psnr
+
+
+def _gaussian_kernel(channel, kernel_size, sigma, device):
+    def gaussian(kernel_size, sigma, device):
+        gauss = torch.arange(
+            start=(1 - kernel_size) / 2, end=(1 + kernel_size) / 2, step=1, dtype=torch.float32, device=device
+        )
+        gauss = torch.exp(-gauss.pow(2) / (2 * pow(sigma, 2)))
+        return (gauss / gauss.sum()).unsqueeze(dim=0)  # (1, kernel_size)
+
+    gaussian_kernel_x = gaussian(kernel_size[0], sigma[0], device)
+    gaussian_kernel_y = gaussian(kernel_size[1], sigma[1], device)
+    kernel = torch.matmul(gaussian_kernel_x.t(), gaussian_kernel_y)  # (kernel_size, 1) * (1, kernel_size)
+
+    return kernel.expand(channel, 1, kernel_size[0], kernel_size[1])
+
+
+def ssim(
+    pred: torch.Tensor,
+    target: torch.Tensor,
+    kernel_size: Sequence[int] = (11, 11),
+    sigma: Sequence[float] = (1.5, 1.5),
+    reduction: str = "elementwise_mean",
+    data_range: float = None,
+    k1: float = 0.01,
+    k2: float = 0.03
+) -> torch.Tensor:
+    """
+    Computes Structual Similarity Index Measure
+
+    Args:
+        pred: Estimated image
+        target: Ground truth image
+        kernel_size: Size of the gaussian kernel. Default: (11, 11)
+        sigma: Standard deviation of the gaussian kernel. Default: (1.5, 1.5)
+        reduction: A method for reducing ssim over all elements in the ``pred`` tensor. Default: ``elementwise_mean``
+
+            Available reduction methods:
+            - elementwise_mean: takes the mean
+            - none: pass away
+            - sum: add elements
+
+        data_range: Range of the image. If ``None``, it is determined from the image (max - min)
+        k1: Parameter of SSIM. Default: 0.01
+        k2: Parameter of SSIM. Default: 0.03
+
+    Returns:
+        A Tensor with SSIM
+
+    Example:
+
+        >>> pred = torch.rand([16, 1, 16, 16])
+        >>> target = pred * 1.25
+        >>> ssim(pred, target)
+        tensor(0.9520)
+    """
+
+    if pred.dtype != target.dtype:
```

Comment:
FYI, all of these if clauses lead to a decrease in converage since tests don't hit them. If we want to keep coverage up, we need to have tests for these cases.

Reply:
Can you share me an example of testing for that as I haven't done tests for that before?

Reply:
Here is a random example of how you can test for raised exceptions
https://github.com/PyTorchLightning/pytorch-lightning/blob/25ee51bc570503f331dceecc610d0eb355e22327/tests/callbacks/test_progress_bar.py#L52
You just feed the wrong inputs to your metric and check that the right error message is raised.

Reply:
Thank you.